ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -0,0 +1,356 @@
+/// This adds another >= log N bits, where N is the amount of elements
+///
+/// This implementation of quicksort is in place and uses a stack instead of recursion
+/// It terminates once the stack is empty
+/// # Errors
+/// Will propagate errors from transport and a few typecasts
+pub async fn quicksort_by_key_insecure<C, K, F, S>(
+    ctx: C,
+    list: &mut [S],
+    desc: bool,
+    f: F,
+) -> Result<(), Error>
+where
+    C: Context,
+    S: Send + Sync,
+    F: Fn(&S) -> &AdditiveShare<K> + Send + Sync,
+    for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
+    K: WeakSharedValue + Field + CustomArray<Element = Boolean>,
+{
+    // create stack
+    let mut stack: Vec<(C, usize, usize)> = vec![];
+
+    // initialize stack
+    stack.push((ctx, 0usize, list.len()));
+
+    // iterate through quicksort recursions
+    while let Some((ctx, b_l, b_r)) = stack.pop() {
+        // start of quicksort function
+        // check whether sort is needed
+        if b_l + 1 < b_r {
+            // set up iterator
+            let mut iterator = list[b_l..b_r].iter();
+            // first element is pivot, apply key extraction function f
+            let pivot = f(iterator.next().unwrap());
+            // create pointer to context for moving into closure
+            let pctx = &ctx;
+            let pf = &f;
+            // precompute comparison against pivot and reveal result in parallel
+            let comp = seq_join(
+                ctx.active_work(),
+                stream_iter(iterator.enumerate().map(|(n, x)| {
+                    async move {
+                        // compare current element against pivot
+                        let sh_comp = compare_gt(
+                            pctx.narrow(&Step::Compare)
+                                .set_total_records(b_r - (b_l + 1)),
+                            RecordId::from(n),
+                            // apply key extraction function f to element x
+                            pf(x),
+                            pivot,
+                        )
+                        .await?;
+
+                        // reveal outcome of comparison
+                        sh_comp
+                            .reveal(
+                                pctx.narrow(&Step::Reveal)
+                                    .set_total_records(b_r - (b_l + 1)),
+                                RecordId::from(n),
+                            )
+                            .await
+                    }
+                })),
+            )
+            .try_collect::<Vec<_>>()
+            .await?;
+
+            // swap elements based on comparisons
+            // i is index of first element larger than pivot
+            let mut i = b_l + 1;
+            for j in b_l + 1..b_r {
+                // desc = true will flip the order
+                if comp[j - (b_l + 1)] == Boolean::from(false ^ desc) {
+                    list.swap(i, j);
```

Comment:
This is a bit confusing since I wanted to do all comparisons in parallel and then in a second step reorder the elements based on the comparison. Element list[b_l] is the pivot element and comp[b_l] stores the outcome of the comparison of list[b_l+1] with the pivot. comp has only length b_r-(b_l+1) rather than b_r-b_l since we do not include the pivot element (and don't compare the pivot to itself).

Reply:
Still, the first time through this loop, this code potentially calls:
```
list.swap(1, 1);
```

I assume that's an optimized case and nothing actually happens (I hope), and we need to increment `i` in that case, so there isn't an obvious way to avoid this.